ext/options.go

```diff
@@ -0,0 +1,50 @@
+package ext
+
+import (
+	"github.com/opentracing/opentracing-go"
+)
+
+// TODO move RPCServerOption into this file? OR, remove it. Code such as:
+//
+// err := tracer.Extract(...)
+// if err != nil { ... }
+// tracer.StartSpan("...", RPCServerOption(...))
+//
+// can be rewritten as:
+// tracer.StartSpan("...", SpanKindRPCServer, ChildOfCarrierOption(...))
+
+type extractedOption struct {
+	ref             opentracing.SpanReferenceType
+	format, carrier interface{}
+}
+
+// Apply implements the StartSpanOption interface.
+func (e extractedOption) Apply(tracer opentracing.Tracer, o *opentracing.StartSpanOptions) {
+	spanCtx, err := tracer.Extract(e.format, e.carrier)
+	if spanCtx != nil {
+		opentracing.SpanReference{e.ref, spanCtx}.Apply(tracer, o)
+	}
+	if err != nil {
+		opentracing.Tag{string(CarriedContextError), err.Error()}.Apply(tracer, o)
+	}
+}
+
+// CarriedSpanRefOption extracts a context from "carrier" (of type "format")
+// and returns a StartSpanOption that will the associated context as a span
+// reference of type "ref" to a newly started span.
+func CarriedSpanRefOption(ref opentracing.SpanReferenceType,
+	format, carrier interface{}) opentracing.StartSpanOption {
+	return extractedOption{ref, format, carrier}
+}
+
+// CarriedChildOfOption constructs a opentracing.ChildOfRef SpanReference option
+// for use in StartSpan with a carried context.
+func CarriedChildOfOption(format, carrier interface{}) opentracing.StartSpanOption {
```

Comment:
this would arguably flow better in calling code as

```
tracer.StartSpan("...", ext.ChildOfCarrier(format, carrier))
```

(i.e., get rid of the Option suffix... it's still obviously a StartSpanOption thanks to the return value if we're worried about greppability)